[CZBANK-56] feature: Implement better-auth to handle auth

vojtech-cerveny opened the pull request:
Sorry for the big code dump — but this is a full refactor of our auth flow.

We’re introducing **better-auth**, which claims to be the most comprehensive authentication framework for TypeScript (their words 😄). It gives us more flexibility and opens up options like API rate limits, OAuth, role management, and more.

I’ve tried to ensure all API endpoints behave consistently (or at least similarly). The **user schema** has changed — this is all documented in the OpenAPI spec.

## Breaking Changes
- **API Keys** — The key is now visible **only** at the moment of creation. After that, it’s hashed and can’t be retrieved.  
  This also means we can’t directly insert API keys into the database (e.g., when seeding users) as we did before with static values like `33`.
---

## What I’d Like to Discuss
- How should we handle **client-side better-auth requests**?  
  Should we explicitly mark them in `user-service.ts` (e.g., as “client-side only”)?  
  Or should we just keep them in the `page.tsx` (or wherever they’re used)?  
  I’m unsure — would love to hear your thoughts!

---

## How to test it
- It can be tricky - maybe you need to wipe your DB, because it changes schema a lot! And I didn't want to play with migrations to create easy way how to go up and down with migrations.

## Screenshots

<img width="3355" height="3033" alt="image" src="https://github.com/user-attachments/assets/806cf9a9-4bd4-49af-be04-ab2808c95611" />
<img width="552" height="434" alt="Screenshot 2025-07-21 at 23 22 05" src="https://github.com/user-attachments/assets/cf88bb37-7490-4511-a175-92c4ed864a1e" />
<img width="1144" height="744" alt="image" src="https://github.com/user-attachments/assets/e762c8e8-3f16-46e3-983f-51a343399b62" />

postulkam committed:
feat: added button for admin section into user ctx menu

vojtech-cerveny committed:
refactor: integrate better-auth for user authentication and update user model

- Replaced NextAuth with better-auth for user authentication, including session management and user creation.
- Updated the User model in Prisma schema to include new fields such as emailVerified, image, and role.
- Added new Session and Account models to support better-auth functionality.
- Refactored user service to utilize better-auth for user registration and sign-in processes.
- Adjusted various imports and function calls across the application to align with the new authentication flow.

vojtech-cerveny committed:
feat: add API key management and user account enhancements

- Introduced API key creation and management functionality, allowing users to generate and delete API keys.
- Updated the User model in Prisma schema to include a new Apikey model for storing API keys.
- Enhanced user registration to automatically create a default API key upon user creation.
- Refactored authentication flow to utilize better-auth for session management and API key operations.
- Improved user profile page to display and manage API keys effectively.
- Added new UI components for API key creation and display, including copy and delete functionalities.

vojtech-cerveny committed:
feat: enhance authentication and user management features

- Implemented API key rate limiting in the auth configuration for better control over API usage.
- Updated the user session retrieval process to utilize better-auth's API for session management.
- Refactored the UserButton component to separate client and server logic, improving maintainability.
- Enhanced the registration form to provide user feedback upon successful account creation.
- Adjusted user data retrieval to include bank accounts in the user service, streamlining user information access.
- Improved transaction components to correctly display user avatars and names based on updated data structure.

vojtech-cerveny committed:
refactor: consolidate UserButton component logic and improve session handling

- Removed the UserButtonClient component and integrated its functionality directly into the UserButton component for better maintainability.
- Updated the UserButton to utilize the useSession hook for session management, enhancing the user experience with loading states and redirection to the sign-in page.
- Simplified the rendering logic for user authentication states, ensuring a more streamlined user interface.

vojtech-cerveny committed:
docs: add useful links section to README

- Added a new section for useful links in the README to provide additional resources for better-auth, Prisma, and Next.js integration.

vojtech-cerveny committed:
feat: enhance user authentication and registration flow

- Updated user creation process to utilize better-auth's API for improved user management.
- Introduced a new constants file to define minimum password length, enhancing password validation across the application.
- Refactored the sign-in and registration forms to dynamically display the minimum password length.
- Added a new API route for listing API keys, improving user access to their API management.
- Removed deprecated bank account retrieval function to streamline the codebase.

vojtech-cerveny committed:
refactor: streamline user seeding process and update bank account handling

- Refactored the user seeding logic to utilize a more structured approach with a dedicated `seedUsers` function.
- Updated the bank account handling to reflect changes in the user data structure, ensuring accurate account number updates.
- Removed deprecated user data structures and improved logging for better traceability during the seeding process.
- Enhanced the main function to clean up the database before seeding and generate transactions, maintaining existing logic.

vojtech-cerveny committed:
refactor: simplify API key display and remove unused copy functionality

- Removed the handleCopyKey function to streamline the API key form component.
- Updated the API key display to show only the start of the key for improved security.
- Enhanced the header to handle cases where the API key name may be undefined, providing a default label.

vojtech-cerveny committed:
chore: nemae ApiKeyCard component for improved API key management

- Replaced ApiKeyForm with the new ApiKeyCard component in the profile page for better display and management of API keys.
- Added ApiKeyCard component to handle API key details, including expiration status and delete functionality.
- Created CreateApiKey component for generating new API keys, enhancing user experience in API key management.

vojtech-cerveny committed:
refactor: centralize rate limit configuration and clean up user repository

- Introduced a new constants file to define rate limit settings for API usage, enhancing maintainability.
- Updated the auth configuration to utilize the centralized rate limit constants.
- Removed deprecated API key regeneration logic from the user repository to streamline the codebase.
- Added a new function in the user repository to fetch all users along with their bank accounts, improving data retrieval efficiency.

vojtech-cerveny committed:
feat: add session management to authentication configuration

- Introduced session management settings in the auth configuration, specifying expiration and update age for user sessions.
- Added a new SESSION constant in the constants file to define session parameters, enhancing the overall authentication flow.

vojtech-cerveny committed:
chore: add CI configuration for Docker and update Dockerfile (#5)

vojtech-cerveny committed:
chore: refactor prisma-migrate service in CI configuration

- Updated the prisma-migrate service in docker-compose.ci.yml to build from a new Dockerfile.migration instead of using a pre-defined image.
- Added Dockerfile.migration to handle dependency installation and migration commands for Prisma.

vojtech-cerveny committed:
chore: simplify dependency declaration for prisma-migrate in CI configuration

- Updated the docker-compose.ci.yml to streamline the dependency declaration for the prisma-migrate service by removing the condition and using a simpler format.

vojtech-cerveny committed:
chore: remove unnecessary port mapping from CI configuration

- Updated docker-compose.ci.yml to remove the port mapping for the service, as it is not needed in the CI environment.

vojtech-cerveny committed:
chore: update CI configuration to prevent service restart

- Added a restart policy set to "no" for the service in docker-compose.ci.yml to ensure it does not restart automatically in the CI environment.

vojtech-cerveny committed:
chore: add volume configuration for PostgreSQL data in CI setup

- Added a volume declaration for czechibank-db-data in docker-compose.ci.yml to persist PostgreSQL data across container restarts.

postulkam commented on src/domain/user-domain/user-service.ts:
```diff
@@ -0,0 +1,144 @@
+  /**
+   * Creates a user using better-auth and initializes a bank account.
+   */
+  async createUserFromEmail(userData: { email: string; name: string; password: string }, isAPI: boolean = false) {
```
Instead of 
`userData: { email: string; name: string; password: string }`
it will be better and more useful to use created interface or use the Zod schema here - that will be really cool.


```
export const CreateUserDataSchema = z.object({
  email: z.string().email(),
  name: z.string().min(1),
  password: z.string().min(6),
});

export type CreateUserData = z.infer<typeof CreateUserDataSchema>;
```

Or old good interface

```
export interface CreateUserData {
  email: string;
  name: string;
  password: string;
}
```

With this, we can use this type in all function calls and the IDE will automatically underline the problem (you can try it, add something to `user-actions.ts` function `processUserRegistration` => the IDE will not have any problems with the added parameters and that is not right).

```
  const userData = {
    email: formData.get("email") as string,
    name: formData.get("name") as string,
    password: formData.get("password") as string,
    dva: 'kkk'
  };
  ```

Try `satisfies` here with the corresponding interface and you will see the usefulness of TypeScript. :)

```
  const userData = {
    email: formData.get("email") as string,
    name: formData.get("name") as string,
    password: formData.get("password") as string,
    dva: 'kkk'
  } satisfies CreateUserData
  ```

vojtech-cerveny replied:
Fixed, thanks for suggestion.

postulkam commented on src/app/page.tsx:
```diff
@@ -1,12 +1,15 @@
-  const session = await getSession();
+  const session = await auth.api.getSession({
+    headers: await headers(),
+  });
```
I think, all of these `auth.api.getSession...` should be centralized in one file that will serveas the only access point for `auth.api`.
We discussed it with prisma on call, this is similar principle - there should be one access point for database operations / auth operations in app.

There are more advantages of this solution
- we can handle some redirect in one place
- we can react to error codes in one place
- no duplicated code
- better refactoring in future

postulkam commented on src/components/auth/user-button.tsx:
```diff
@@ -5,30 +5,26 @@ import {
+  const session = await auth.api.getSession({
+    headers: await headers(),
+  });
+  if (!session?.user?.id) {
```
Similar to previous comment about centralizing...

This is perfect example - in `app/page.tsx` we are checking only the `!session` and redirecting. Here we are checking `!session?.user?.id` and that is confusing.

vojtech-cerveny replied:
Yeah, this is something we should cover by moving this stuff into userService.

postulkam commented on src/domain/user-domain/user-service.ts:
```diff
@@ -0,0 +1,137 @@
+          toast({
+            title: "Error",
+            description: ctx.error.message,
+            variant: "destructive",
```
Just note for the future - there are more ways of some toast, messages, whatever. We should use only one principle in whole app, this is confusing for developers and users.

vojtech-cerveny replied:
I will create ticket for it - to create `createToast()` function to have one source of truth. https://czechitas-jira.atlassian.net/browse/CZBANK-60

vojtech-cerveny committed:
chore: add database setup script and update package.json

- Introduced a new script `setup-database.ts` for setting up the database, including generating the Prisma client, running migrations, and seeding the database if applicable.
- Updated `package.json` to include a new command `db:ci:setup` for executing the database setup script.

vojtech-cerveny committed:
test: if preview apps works properly or not.

postulkam commented on scripts/seed.ts:
```diff
@@ -1,215 +1,122 @@
+          email: userSeed.email,
+          name: userSeed.name,
+          password: userSeed.password,
+          role: userSeed.role as "user" | "admin",
```
Better use some enum for "user" | "admin" or look inside the Auth library, there will be enum for export, maybe.

vojtech-cerveny replied:
Exporting `export type Role = "admin" | "user";` from `"@/lib/permissions";` - wdyt?

postulkam commented on scripts/seed.ts:
```diff
@@ -1,215 +1,122 @@
 
-// Add lektor users to the main users array
-users.push(...lektorUsers);
+      const user = response.user;
```
How are we handling errors here? Are we sure that it always returns the response with the user data?

vojtech-cerveny replied:
Documentation saing this: https://www.better-auth.com/docs/concepts/api#error-handling - so only `try/catch`, with their `ApiError` instance. 

When this is seed it should fail if anything is wrong, I am OK to have big try/catch, but I am open to suggestiong.

postulkam replied:
Big try/catch is ok for this situation of creating testing data. ;)

postulkam commented on scripts/seed.ts:
```diff
@@ -1,215 +1,122 @@
+      // Update API key if deterministic value is needed
+      // TODO: @vojtech-cerveny - this is not working due the api keys are hashed in the DB, so we are not able to update them directly.
+      // and API-keys are generated on the fly, so we don't know the values - just after creation.
+      await prisma.apikey.update({
```
Should this code run even if the bankAccount is non-existent? The if/else will handle the non-existence, but the code will run and update the `apiKey` no matter of the `bankAccount` existence.

vojtech-cerveny replied:
Well, right now, we have hook for creating bankAccount after creating account. So user should have always bankAccount. And API key is not connected to bankAccount, only to account (user).

postulkam commented on scripts/seed.ts:
```diff
@@ -308,12 +215,12 @@ function generateTransactionData(regularUsers: any[], totalTransactions: number)
+        senderBalance: sender.bankAccounts[0].balance - amount,
+        receiverBalance: receiver.bankAccounts[0].balance + amount,
+        senderAccountId: sender.bankAccounts[0].id,
+        receiverAccountId: receiver.bankAccounts[0].id,
```
You are repeatedly accessing the `sender.bankAccounts[0]` and `receiver.bankAccounts[0]`

Normally, we will create constant before that and use it.

```
const receiverAccount = receiver.bankAccounts[0]
const senderAccount = sender.bankAccounts[0]
....

Btw, the `0` is right? Woudn't be the `offset` from function parameters better?

postulkam commented on src/lib/auth-client.ts:
```diff
@@ -0,0 +1,19 @@
+import { ac, admin, user } from "./permissions";
+
+export const authClient = createAuthClient({
+  baseURL: "http://localhost:3000", // The base URL of your auth server
```
Is the `localhost` here ok? Is this used only for development?

vojtech-cerveny replied:
Nice catch - I checked documentation and we can remove it - https://www.better-auth.com/docs/installation#create-client-instance

*(If the auth server is running on the same domain as your client, you can skip this step.)*

postulkam commented on src/app/api/v1/server-actions.ts:
```diff
@@ -1,25 +1,40 @@
+import { auth } from "../../../../auth";
 
-export async function checkUserAuthOrThrowError(request: Request): Promise<User | ErrorResponse> {
+export async function checkUserAuthOrThrowError(request: Request): Promise<any | ErrorResponse> {
```
Why return `Promise<any | ...>` and not `User` from Prisma?

vojtech-cerveny replied:
fixed by `typeof auth.$Infer.Session.user` - I want to stick better-auth for users types, wdyt? https://www.better-auth.com/docs/concepts/typescript#inferring-types

postulkam replied:
I will deep dive into these types in the near future, `auth.$Infer.Session.user` is ok, but it is too long and that always sound "sketchy". :D

postulkam commented on src/app/api/v1/server-actions.ts:
```diff
@@ -1,25 +1,40 @@
+  console.log("apiKeyResponse", apiKeyResponse);
+  if (!apiKeyResponse.valid) {
+    return errorResponse(apiKeyResponse.error?.message ?? "Invalid API KEY", ApiErrorCode.OPERATION_FAILED);
+  }
```
Should we redirect or logout the user with non valid apiKey? The same applies for the session below.

vojtech-cerveny replied:
This is the open API, so we this is for direct usage of API - i think that we don't have options to redirect or logout - I guess it has some session, but I don't think that will be our problem. If you have incorrect API key, we should just inform user about it. That it is already expired or it is wrong. WDYT?

vojtech-cerveny replied:
This server-actions just for API. And for Session below - I don't think that we will see error there in future. If your API key is valid, then you we are creating session - but just for get correct user from session. And it should be always this flow.

postulkam replied:
I agree that we should inform the user about the expiration of non-existence of apiKey.

postulkam commented on src/app/api/v1/user/create/route.ts:
```diff
@@ -54,17 +55,30 @@ export async function POST(request: Request) {
+
+    return Response.json({ ...response.user, apiKey: apiKey.key }, { status: 201 });
   } catch (error) {
+    // TODO: @vojtech-cerveny - handle better-auth - we can reuse their ApiErrorCodes etc.
```
Yes! Use the ApiErrorCodes :)

postulkam commented on src/app/bankAccount/[id]/page.tsx:
```diff
@@ -75,15 +79,17 @@ export default async function BankAccountPage(props: { params: Promise<{ id: str
+                  userId={session.user.id}
+                  allUsers={allUsers.data.map((user) => ({
+                    ...user,
+                    BankAccount: user.bankAccounts,
```
Why the spread operator for `...user` here? 
Is it because of the `BankAccount` with the uppercase? Why are we using the uppercase "B" only in this parameter?

vojtech-cerveny replied:
Fixed! Good catch, changed to `allUsers={allUsers.data}`

postulkam commented on src/app/profile/page.client.tsx:
```diff
@@ -0,0 +1,163 @@
+
+      await authClient.updateUser(
+        {
+          image: JSON.stringify(avatarConfig),
```
I think that this solved Silvia in her PR - there is double `stringify`.

postulkam commented on src/app/profile/page.client.tsx:
```diff
@@ -0,0 +1,163 @@
+    } finally {
+      setIsUpdatingAvatar(false);
+    }
+  };
```
Is this `catch` and `finally` really necessary? There is already `onSuccess` and `onError` function, it seems like duplication of some logic for me.

vojtech-cerveny replied:
Good catch!

postulkam commented on src/app/profile/page.client.tsx:
```diff
@@ -0,0 +1,163 @@
+            <CreateApiKey />
+          </div>
+
+          {apiKeys.length === 0 ? (
```
Curious question - is it possible to create user without apiKey?

vojtech-cerveny replied:
Yes, it is. If you creating user via FE - with new flow of creating and seeing API keys, we are not able to create api key during creating user (he would be not be able to use API key). So for newly created users (via FE, not API) there is no API key and user needs to create it manually (or via API)

postulkam commented on src/components/profile/apikey-card.tsx:
```diff
@@ -0,0 +1,106 @@
+        },
+      },
+    );
+  };
```
I really like these `onSuccess` and `onError` blocks in auth calls.

postulkam commented on src/components/profile/create-api-key.tsx:
```diff
@@ -0,0 +1,159 @@
+
+export default function CreateApiKey() {
+  const router = useRouter();
+  const [isLoading, setIsLoading] = useState(false);
```
Consider to make it default as true - it will always be loading at the start or not?

vojtech-cerveny replied:
The state variable (isLoading) represents whether the API key creation process is currently happening. I will rename it to "isCreatingApiKey" to be more clear what it store. WDYT?

postulkam replied:
Definitely, `isCreatingApiKey` sounds better!

postulkam commented on src/components/register/form.tsx:
```diff
@@ -35,13 +39,33 @@ export function RegisterForm() {
+          });
+        },
+      },
+    );
```
Just one thing - sometimes you use the `onSucces` and `onError` with combination of `await` and sometimes not. Can you say why?

vojtech-cerveny replied:
Yeah, probably I forgot to use it somewhere. This is why 😅 I added it here and I will create new commit with generating it in userService.

postulkam reviewed (commented):
That was... so much, but GREAT JOB! Really!

Don't be afraid of so many comments, some of them are just my curious questions, lots of them are some typescript suggestions and it is up to you if you want to implement them or not. 
We can discuss it on call or maybe do some pair-programming if you'd like. :)

vojtech-cerveny committed:
style: :art: npx prisma format

vojtech-cerveny committed:
refactor: better types

vojtech-cerveny committed:
refactor: removing unnecessary code / renaming

- try/catch blocks
- renaming isLoading -> isCreating
- stop using sex property for user for now.
- removing baseUrl from authClient - it was unncessary in our case
- remove old auth.ts file - we don't use it anymore

vojtech-cerveny committed:
chore: update seed command in database setup script

- Changed the seed command from `npx prisma db seed` to `pnpm db:seed` in the `setup-database.ts`

vojtech-cerveny commented:
One thing to discuss - I think that we should split userService into two pieces - one for client, second for server.

```typescript

const userService {
  client: {
    async signIn(ctx){
      authClient.singIn(something)
    },
    async singOut(){
    }
  }
  // most of this would be for API I guess (at least here in userService)
  server: {
    async createUser(){
      auth.api.createUser()
    }
  }	
}
```

WDYT?

vojtech-cerveny committed:
refactor: replace authClient with userService and apikeyService

- Updated the authentication and API key management logic to utilize userService and apikeyService for better organization and maintainability.
- Removed direct dependencies on authClient in various components and routes, streamlining the codebase.
- Enhanced session management by consistently using userService for session retrieval across different pages.

vojtech-cerveny committed:
fix: fix some wrong Currency import and extract value

postulkam committed:
feat: admin page + admin button in user ctx menu

postulkam commented on src/components/register/form.tsx:
```diff
@@ -130,7 +146,7 @@ export function RegisterForm() {
             </FormItem>
           )}
-        />
+        /> */}
```
Better create an issue in JIRA, we will forget about this. ;)

vojtech-cerveny replied:
Yeah, you are right. If we need that, we will implement it. Right now, it is useless.

postulkam commented on src/domain/apikey/apikey-service.ts:
```diff
@@ -0,0 +1,86 @@
+          error: {
+            message: parsed.error.message,
+          },
+        } as ErrorContext);
```
Will this error automatically appear in some toast for user?

vojtech-cerveny replied:
Yeah, it is in `create-api-key.tsx`

```ts
    await apikeyService.client.createApiKey(
      {
        name: data.name,
        expiresIn: data.expiresIn,
      },
      {
        onSuccess: (context) => {
          const apiKey = context.data as Apikey;
          setNewApiKey(apiKey.key);
          toast({
            title: "API key created",
            description: "Your new API key has been created successfully",
          });
          router.refresh();
          setInCreating(false);
        },
        onError: (error) => {
          toast({
            description: `Failed to create API key due to an error: ${error?.error?.message}`,
            variant: "destructive",
            title: "Error",
          });
          setInCreating(false);
        },
      },
    );
  };

```

postulkam commented on src/domain/user-domain/user-schema.ts:
```diff
@@ -1,8 +1,15 @@
+
+export const LoginSchema = z.object({
+  email: z.string().email(),
+  password: z.string().min(MIN_PASSWORD_LENGTH, `Password must be at least ${MIN_PASSWORD_LENGTH} characters long`),
```
You can use `extends` of the `LoginSchema` here => in that way, you won't need to specify shared parameters again.

```
const UserBaseSchema = z.object({
  email: z.string().email(),
  password: z.string().min(MIN_PASSWORD_LENGTH, `Password must be at least ${MIN_PASSWORD_LENGTH} characters long`),
});

export const LoginSchema = UserBaseSchema;

export const UserSchema = UserBaseSchema.extend({
  name: z.string().trim().min(1, "Name cannot be empty or contain only spaces"),
});

export type CreateUserSchema = z.infer<typeof UserSchema>;
```

vojtech-cerveny replied:
Fixed.

postulkam reviewed (commented):
Some small suggestions.

I made a commit with an Admin page - can you look at it please? :)

postulkam commented:
> One thing to discuss - I think that we should split userService into two pieces - one for client, second for server.
> 
> ```ts
> const userService {
>   client: {
>     async signIn(ctx){
>       authClient.singIn(something)
>     },
>     async singOut(){
>     }
>   }
>   // most of this would be for API I guess (at least here in userService)
>   server: {
>     async createUser(){
>       auth.api.createUser()
>     }
>   }	
> }
> ```
> 
> WDYT?


I have seen it during the review and I like it.
Only thing that I don't really like - there are functions that are named as the same in client x server part of code. But I understand why and we can keep it like that for now.

----

Other questions - I noticed that you have another PRs here and you made lots of changes there. How do you plan to merge them and this together? I think it will be easier to merge this PR first and then rebase/cherry-pick one commit after another from your other PRs and I am rooting for you, it will not be pleasant...

vojtech-cerveny committed:
chore: remove sex field from schema and seed data

- Deleted the Sex enum from schema.prisma.
- Removed sex property from user seed data in seed.ts.
- Commented out the sex field in the registration form to address current authentication issues.
- Updated Swagger documentation to reflect the removal of the sex field.

vojtech-cerveny committed:
refactor: consolidate user schema definitions

- Introduced UserBaseSchema to reduce redundancy in user schema definitions.
- Updated LoginSchema to extend UserBaseSchema for consistency.
- Adjusted CreateUserSchema type inference to reflect changes in UserSchema.
- Modified onError parameter type in user service for better error handling.

vojtech-cerveny committed:
Merge branch 'develop' into CZBANK-56-better-auth

vojtech-cerveny commented:
> How do you plan to merge them and this together? I think it will be easier to merge this PR first and then rebase/cherry-pick one commit after another from your other PRs and I am rooting for you, it will not be pleasant...

👍 Yeah, this is biggest PR, we should merge it first. Then I will fight with git and try to rebase other PRS.

vojtech-cerveny committed:
Merge branch 'develop' into CZBANK-56-better-auth

vojtech-cerveny committed:
fix: ensure avatar update state is reset on success and error

- Added `setIsUpdatingAvatar(false)` to both the success and error handlers of the avatar update function to properly reset the updating state after the operation completes.

postulkam reviewed (approved):
LGTM
